imp_developer.py

```diff
@@ -272,56 +326,414 @@ def __base64_encode(str):
+            headers=HttpHeaders.AUTH)
+        error = ImpCentral.handle_http_response(response, code)
+        return response, error
+
+    @staticmethod
+    def account(token):
+        response, code = HTTP.get(token,
+            PL_IMPCENTRAL_API_URL_V5 + "/accounts/me")
+        error = ImpCentral.handle_http_response(response, code)
+        return response.get("data"), error
+
+
+    @staticmethod
+    def list_products(token, owner_id=None):
+        ImpCentral.account(token)
+        # TODO: list all products
+        if owner_id:
+            data='{"filter[owner.id]:"' + owner_id + '"}'
+        else:
+            data=None
+
+        link = PL_IMPCENTRAL_API_URL_V5 + "products"
+        products = []
+
+        while link != None:
+            response, code = HTTP.get(token, link, data=data)
+            error = ImpCentral.handle_http_response(response, code)
+            data = None
+
+            if not error:
+                for product in response.get("data"):
+                    if product["relationships"]["owner"]["id"] == owner_id:
+                        products.append(product)
+                next_link = response["links"].get("next")
+                if next_link == link:
+                    link = None
+                else:
+                    link = next_link
+            else:
+                return response, error
+
+        return products, error
+
+    @staticmethod
+    def list_devicegroups(token, product_id):
+        devicegroups = []
+        link = PL_IMPCENTRAL_API_URL_V5 + "devicegroups"
+        data = '{"filter[product.id]:"' + product_id + '"}'
+        error = None
+
+        while link != None:
+            response, code = HTTP.get(token, link, data=data)
+            error = ImpCentral.handle_http_response(response, code)
+            data = None
+
+            if not error:
+                for devicegroup in response.get("data"):
+                    if devicegroup["relationships"]["product"]["id"] == product_id:
+                        devicegroups.append(devicegroup)
+                    next_link = response["links"].get("next")
+                    if next_link == link:
+                        link = None
+                    else:
+                        link = next_link
+            else:
+                return response, error
+
+        return devicegroups, error
+
+    @staticmethod
+    def list_devices(token, devicegroup_id=None):
+        link = PL_IMPCENTRAL_API_URL_V5 + "devices"
+        devices = []
+        # filter by group id or not
+        if devicegroup_id:
+            data='{"filter[devicegroup.id]": "' + devicegroup_id + '"}'
+        else:
+            data=None
+
+        while link != None:
+            response, code = HTTP.get(token, url=link, data=data)
+            error = ImpCentral.handle_http_response(response, code)
+            data = None
+            if not error:
+                for device in response.get('data'):
```

Comment:
Looks very similar to what we have above... Can we refactor this to avoid the code duplication?

Reply:
reworked